src/main/scala/Validator.scala

```diff
@@ -76,22 +147,170 @@ class Validator(var tableCsvFile: URI, sourceUri: String = "") {
+  }
+
+  private def validatePrimaryKeys(
+      allPrimaryKeyValues: mutable.Set[List[Any]],
+      row: CSVRecord,
+      validateRowOutput: ValidateRowOutput
+  ) = {
+    val primaryKeyValues = validateRowOutput.primaryKeyValues
+    if (
+      primaryKeyValues.nonEmpty && allPrimaryKeyValues.contains(
+        primaryKeyValues
+      )
+    ) {
+      errors = errors :+ ErrorWithCsvContext(
+        "duplicate_key",
+        "schema",
+        row.toString,
+        "",
+        s"key already present - ${fetchPrimaryKeyString(primaryKeyValues)}",
+        ""
+      )
+    } else allPrimaryKeyValues += primaryKeyValues
+  }
+
+  private def setParentTableForeignKeyReferences(
+      validateRowOutput: ValidateRowOutput,
+      parentTableForeignKeyReferences: Map[ParentTableForeignKeyReference, Set[
+        KeyWithContext
+      ]]
+  ) = {
+    validateRowOutput.parentTableForeignKeyReferences
+      .foreach {
+        case (k, value) => {
+          val allPossibleParentKeyValues =
+            parentTableForeignKeyReferences.getOrElse(
+              k,
+              Set[KeyWithContext]()
+            )
+          if (allPossibleParentKeyValues.contains(value)) {
+            allPossibleParentKeyValues -= value
+            value.isDuplicate = true
+          }
+          allPossibleParentKeyValues += value
+          parentTableForeignKeyReferences(k) = allPossibleParentKeyValues
+        }
+      }
+  }
+
+  private def setChildTableForeignKeys(
+      validateRowOutput: ValidateRowOutput,
+      childTableForeignKeys: Map[ChildTableForeignKey, Set[KeyWithContext]]
+  ) = {
+    validateRowOutput.childTableForeignKeys
+      .foreach {
+        case (k, value) => {
+          val childKeyValues = childTableForeignKeys.getOrElse(
+            k,
+            Set[KeyWithContext]()
+          )
+          childKeyValues += value
+          childTableForeignKeys(k) = childKeyValues
+        }
+      }
+  }
+
+  private def validateForeignKeyReferences(
+      childTableForeignKeysByTable: Map[
+        Table,
+        Map[ChildTableForeignKey, Set[KeyWithContext]]
+      ],
+      parentTableForeignKeyReferencesByTable: Map[
+        Table,
+        Map[ParentTableForeignKeyReference, Set[KeyWithContext]]
+      ]
+  ) = {
+    // Child Table : Parent Table
+    // Country, Year, Population  : Country, Name
+    // UK, 2021, 67M  : UK, United Kingdom
+    // EU, 2021, 448M : EU, Europe
+    for (
+      (parentTable, mapParentTableForeignKeyReferenceToAllPossibleValues) <-
+        parentTableForeignKeyReferencesByTable
+    ) {
+      for (
+        (parentTableForeignKeyReference, allPossibleParentTableValues) <-
+          mapParentTableForeignKeyReferenceToAllPossibleValues
+      ) {
+        val childTableForeignKeys
+            : Map[ChildTableForeignKey, Set[KeyWithContext]] =
+          childTableForeignKeysByTable.get(
+            parentTableForeignKeyReference.childTable
+          ) match {
```

Comment:
The following is more succinct and should work:
```
val childTableForeignKeys: Map[ChildTableForeignKey, Set[KeyWithContext]] =
  childTableForeignKeysByTable
    .get(parentTableForeignKeyReference.childTable)
    .getOrElse(
      throw new Exception(s"Could not find corresponding child table for parent table ${parentTable.url}")
    )
```

Reply:
Okay